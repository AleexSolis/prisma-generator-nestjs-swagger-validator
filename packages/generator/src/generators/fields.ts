import { JsTypes, annotationDecorators, typeDecorators } from '../constants';
import { isAnnotatedWith } from '../helpers/annotations';
import { DecoratorObject, Field, FieldDtoPayload } from '../types';

const getValidation = (field: Field): DecoratorObject | null => {
  const { type, kind } = field;

  if (kind === 'enum') {
    return {
      decorator: `@IsEnum(${type})`,
      CVImport: 'IsEnum',
      prismaImport: type,
      apiPropertyProps: { enum: type },
    };
  }
  if (kind === 'object') {
    if (field.relationName) return null;
  }

  const decorator = typeDecorators[type as keyof typeof typeDecorators];
  if (!decorator) return null;

  return {
    decorator: `@${decorator}()`,
    CVImport: decorator,
  };
};

export function getField(field: Field): FieldDtoPayload | null {
  const decorators = new Set<string>();
  const classValidatorImports = new Set<string>();
  const prismaImports = new Set<string>();
  const dtoImports = new Set<string>();

  let apiPropertyProps = {};

  if (field.isRequired) {
    decorators.add('@IsNotEmpty()\n');
    decorators.add('@IsDefined()\n');
    classValidatorImports.add('IsNotEmpty');
    classValidatorImports.add('IsDefined');
  } else {
    decorators.add('@IsOptional()\n');
    Object.assign(apiPropertyProps, { required: false });
    classValidatorImports.add('IsOptional');
  }

  const typeDecorator = getValidation(field);

  if (typeDecorator) {
    const {
      decorator,
      CVImport,
      prismaImport,
      dtoImport,
      apiPropertyProps: apiProperties,
    } = typeDecorator;
    decorator && decorators.add(decorator);
    CVImport && classValidatorImports.add(CVImport);
    prismaImport && prismaImports.add(prismaImport);
    dtoImport && dtoImports.add(dtoImport);

    if (apiProperties) {
      Object.assign(apiPropertyProps, apiProperties);
    }
  } else {
    return null;
  }

  annotationDecorators.forEach((decorator) => {
    if (isAnnotatedWith(field, decorator.regexp)) {
      const result = decorator.handler(field);
      if (result.decorator) decorators.add(result.decorator);
      if (result.CVImport) classValidatorImports.add(result.CVImport);

      if (result.apiPropertyProps) {
        Object.assign(apiPropertyProps, result.apiPropertyProps);
      }
    }
  });

  const apiPropertyPropsString = JSON.stringify(apiPropertyProps)?.replace(
    /"enum":"([^"]+)"/g,
    'enum: $1',
  );

  decorators.add(`@ApiProperty(${apiPropertyPropsString})`);

  const stringField = `
    ${[...decorators].join('\n')}${field.name}${
    !field.isRequired ? '?' : ''
  }: ${JsTypes[field.type] || field.type};
  `;

  return {
    stringField,
    classValidatorImports: [...classValidatorImports],
    prismaImports: [...prismaImports],
    dtoImports: [...dtoImports],
  };
}

function getImportStatements(imports: Set<string>, importSource: string) {
  return imports.size > 0
    ? `
      import {
        ${[...imports].filter(Boolean).join(',\n')}
      } from '${importSource}';
      `
    : '';
}

export function getImports(fields: Array<Field>) {
  const prismaImports = new Set<string>();
  const CVImports = new Set<string>(['IsOptional', 'IsInt', 'IsPositive']);
  const dtoImports = new Set<string>();

  fields.forEach((field) => {
    const results = getField(field);
    if (!results) return;

    const {
      prismaImports: pi,
      classValidatorImports: cvi,
      dtoImports: di,
    } = results;

    cvi && cvi.forEach((cv) => CVImports.add(cv));
    pi && pi.forEach((pi) => prismaImports.add(pi));
    di && di.forEach((di) => dtoImports.add(di));
  });

  const CVImportStatements = getImportStatements(CVImports, 'class-validator');
  const prismaImportStatements = getImportStatements(
    prismaImports,
    '@prisma/client',
  );
  const dtoImportStatements =
    dtoImports.size > 0
      ? `import { ${[...dtoImports].join(', ')} } from '.';`
      : '';

  return `
    ${CVImportStatements}
    ${prismaImportStatements}
    ${dtoImportStatements}
  `;
}
