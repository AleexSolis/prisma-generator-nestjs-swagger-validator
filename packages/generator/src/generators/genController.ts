export const genController = (modelName: string): string => {
  const modelNameLower = modelName.toLowerCase();
  return `
        import { Controller, Get, Post, Body, Patch, Param, Delete, Res } from '@nestjs/common';
        import { ApiResponse, ApiTags } from '@nestjs/swagger';
        import { Response } from 'express';
        import { ${modelName}Service } from './${modelName}.service';
        import { Create${modelName}Dto, Update${modelName}Dto, ${modelName}Dto } from './${modelName}.dto';

        @ApiTags('${modelNameLower}')
        @Controller('${modelNameLower}')
        export class ${modelName}Controller {
        constructor(private readonly ${modelNameLower}Service: ${modelName}Service) {}

        @ApiResponse({ type: ${modelName}Dto, status: 201 })
        @Post()
        async create(@Res() res: Response, @Body() create${modelName}Dto: Create${modelName}Dto) {
            const data = this.${modelNameLower}Service.create(create${modelName}Dto);
            res.send(data);
        }

        @ApiResponse({ type: ${modelName}Dto, status: 200, isArray: true })
        @Get()
        find() {
            return this.${modelNameLower}Service.find({});
        }

        @ApiResponse({ type: ${modelName}Dto, status: 200 })
        @Get(':id')
        findOne(@Param('id') id: string) {
            return this.${modelNameLower}Service.findOne({ id: id });
        }

        @ApiResponse({ type: ${modelName}Dto, status: 200 })
        @Patch(':id')
        update(@Param('id') id: string, @Body() update${modelName}Dto: Update${modelName}Dto) {
            return this.${modelNameLower}Service.update({id: id}, update${modelName}Dto);
        }

        @ApiResponse({ type: ${modelName}Dto, status: 200 })
        @Delete(':id')
        delete(@Param('id') id: string) {
            return this.${modelNameLower}Service.delete({id: id});
        }
        }
        `;
};
