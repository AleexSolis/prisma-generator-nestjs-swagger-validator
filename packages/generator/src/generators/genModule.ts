import { DMMF } from '@prisma/generator-helper';
import { camelCase } from '../utils';

export const genModule = (model: DMMF.Model) => {
  //const modelName = camelCase(model.name);

  return `import { Module } from '@nestjs/common'
    import { ${model.name}Service } from './${model.name}.service'
    import { ${model.name}Controller } from './${model.name}.controller'
    
    @Module({
      providers: [${model.name}Service],
      controllers: [${model.name}Controller],
      exports: [${model.name}Service],
    })
    export class ${model.name}Module {}
  `;
};
