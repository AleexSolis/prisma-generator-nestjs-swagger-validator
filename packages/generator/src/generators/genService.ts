export const genService = (modelName: string): string => {
  const modelNameLower = modelName[0].toLowerCase() + modelName.slice(1);

  return `
      import { Injectable } from '@nestjs/common';
      import { Prisma } from '@prisma/client';
      import { PrismaService } from '../prisma/prisma.service';

      @Injectable()
      export class ${modelName}Service{
  
          constructor(private prisma: PrismaService){}
  
          create(data: Prisma.${modelName}CreateInput) {
            return this.prisma.${modelNameLower}.create({ data });
          }
          
          find(args: Prisma.${modelName}FindManyArgs){
            return this.prisma.${modelNameLower}.findMany(args);
          }
  
          findOne(where: Prisma.${modelName}WhereUniqueInput){
            return this.prisma.${modelNameLower}.findUnique({ where });
          }
          
          update(where: Prisma.${modelName}WhereUniqueInput, data: Prisma.${modelName}UpdateInput){
            return this.prisma.${modelNameLower}.update({
              where,
              data,
            });
          }
        
          delete(where: Prisma.${modelName}WhereUniqueInput){
            return this.prisma.${modelNameLower}.delete({ where });
          }
      }
      `;
};
