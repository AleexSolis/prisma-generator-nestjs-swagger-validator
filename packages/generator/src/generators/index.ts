export * from './fields';
export * from './genController';
export * from './genDto';
export * from './genModule';
export * from './genService';
