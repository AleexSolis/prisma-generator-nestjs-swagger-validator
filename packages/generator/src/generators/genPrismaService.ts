import { fileToWrite } from 'src/types';
import path from 'path';

const genPrismaModule = () => {
  return `
  import { Global, Module } from '@nestjs/common';
  import { PrismaService } from './prisma.service';
  
  @Global()
  @Module({
    providers: [PrismaService],
    exports: [PrismaService],
  })
  export class PrismaModule {}
    `;
};

const genPrismaService = () => {
  return `
    import { INestApplication, Injectable, OnModuleInit } from '@nestjs/common';
    import { PrismaClient } from '@prisma/client';

    @Injectable()
    export class PrismaService extends PrismaClient implements OnModuleInit {
        async onModuleInit() {
            await this.$connect();
        }

        async enableShutdownHooks(app: INestApplication) {
            this.$on('beforeExit', async () => {
            await app.close();
            });
        }
    }
    `;
};

export const handleGenPrismaModule = (output: string): fileToWrite[] => {
  return [
    {
      content: genPrismaModule(),
      location: path.join(output, 'prisma', `prisma.module.ts`),
    },
    {
      content: genPrismaService(),
      location: path.join(output, 'prisma', `prisma.service.ts`),
    },
  ];
};
