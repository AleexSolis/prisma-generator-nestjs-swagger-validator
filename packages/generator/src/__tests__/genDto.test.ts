import { getSampleDMMF } from './__fixtures__/getSampleDMMF';
import { genDto } from '../generators';

test('dto generation', async () => {
  const sampleDMMF = await getSampleDMMF();

  sampleDMMF.datamodel.models.forEach((table) => {
    expect(genDto(table)).toMatchSnapshot(table.name);
  });
});
