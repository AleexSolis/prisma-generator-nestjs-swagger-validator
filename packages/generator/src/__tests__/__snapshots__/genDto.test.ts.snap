// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`dto generation: User 1`] = `
"import { ApiProperty, OmitType, PartialType } from '@nestjs/swagger';
import {
        IsNotEmpty,
IsDefined,
IsInt,
IsString,
IsOptional
      } from 'class-validator';
      
      
    
export class UserDto {
@IsNotEmpty()

@IsDefined()

@IsInt()
@ApiProperty({})id: number;

@IsNotEmpty()

@IsDefined()

@IsString()
@ApiProperty({})email: string;

@IsOptional()

@IsString()
@ApiProperty({\\"required\\":false})name?: string;
}

export class CreateUserDto extends OmitType(UserDto, ['id']) {}

export class UpdateUserDto extends PartialType(
  OmitType(UserDto, ['id'])
) {}

class WhereDto{
  
}

class OrderDto{
  
}

export class FilterUserDto {
  @ApiProperty({ required: false })
  where?: WhereDto;

  @ApiProperty({ required: false })
  order?: OrderDto;

  @ApiProperty({ required: false })
  skip?: number;

  @ApiProperty({ required: false })
  take?: number;
}
"
`;

exports[`enum generation: User 1`] = `
"import { ApiProperty, OmitType, PartialType } from '@nestjs/swagger';
import {
        IsNotEmpty,
IsDefined,
IsInt,
IsString,
IsOptional
      } from 'class-validator';
      
      
    
export class UserDto {
@IsNotEmpty()

@IsDefined()

@IsInt()
@ApiProperty({})id: number;

@IsNotEmpty()

@IsDefined()

@IsString()
@ApiProperty({})email: string;

@IsOptional()

@IsString()
@ApiProperty({\\"required\\":false})name?: string;
}

export class CreateUserDto extends OmitType(UserDto, ['id']) {}

export class UpdateUserDto extends PartialType(
  OmitType(UserDto, ['id'])
) {}

class WhereDto{
  
}

class OrderDto{
  
}

export class FilterUserDto {
  @ApiProperty({ required: false })
  where?: WhereDto;

  @ApiProperty({ required: false })
  order?: OrderDto;

  @ApiProperty({ required: false })
  skip?: number;

  @ApiProperty({ required: false })
  take?: number;
}
"
`;
